### PR TITLE
fix: Pasting a dollar amount does not work

### PR DESCRIPTION
**Pull request by nkdengineer**
### Fixed Issues

$ https://github.com/Expensify/App/issues/47270
PROPOSAL:  https://github.com/Expensify/App/issues/47270#issuecomment-2292798433

### Tests

Precondition: Copy an amount (eg: $90.28)

1. Click FAB > Submit expense
2. In the BNP press CTRL+V
3. Verify that: 90.28 should be pasted in and currency will update to USD

- [x] Verify that no errors appear in the JS console

### QA Steps

Precondition: Copy an amount (eg: $90.28)

1. Click FAB > Submit expense
2. In the BNP press CTRL+V
3. Verify that: 90.28 should be pasted in and currency will update to USD
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/327ea480-4208-4d6d-bd8d-0146856d79ae

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/ad8ff44b-b3e9-4de1-9603-e23850594cd5

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c2b8c50c-4c8e-4de0-905d-008b09f6dc56

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/329c145a-f328-4622-b1f5-926576435a9a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/7556ed33-6014-4e18-b575-64c781b655a6

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/8fad14af-2165-4f86-8c19-84be2078f567

</details>

### CODE REVIEW

File: src/components/MoneyRequestAmountInput.tsx [hunk 1]
```diff
@@ -158,14 +163,30 @@ function MoneyRequestAmountInput(
...
+                const updateDecimals = CurrencyUtils.getCurrencyDecimals(updateCurrency);
+                finalAmount = updateAmount.includes('.') ? MoneyRequestUtils.stripCommaFromAmount(updateAmount) : MoneyRequestUtils.replaceCommasWithPeriod(updateAmount);
+                // Use a shallow copy of selection to trigger setSelection
+                // More info: https://github.com/Expensify/App/issues/16385
+                if (!MoneyRequestUtils.validateAmount(finalAmount, updateDecimals)) {
+                    setSelection((prevSelection) => ({...prevSelection}));
+                    return;
+                }
+                setCurrentCurrency(updateCurrency);
+                // onPasteAmountWithCurrency?.(updateCurrency);
```
Comment (s77rt) [inline src/components/MoneyRequestAmountInput.tsx:179]: Let's not repeat the code here. You can use this to get both the amount and the currency, and for code clarity let's always calculate `newDecimals`

```js
const [newCurrency, newAmount] = "$15.77".split(/(\d.*)/, 2)
```

make sure that newAmount is not undefined (in case of pasting just the currency)

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/components/MoneyRequestAmountInput.tsx (hunk 1 above)
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx:179]: @s77rt In case the user only pastes currency without amount, do we update the currency?

File: src/pages/iou/MoneyRequestAmountForm.tsx [hunk 2]
```diff
@@ -57,7 +57,7 @@ type MoneyRequestAmountFormProps = {
     isCurrencyPressable?: boolean;
 
     /** Fired when back button pressed, navigates to currency selection page */
-    onCurrencyButtonPress?: () => void;
+    onCurrencyButtonPress?: (updateCurrency?: string) => void;
```
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/pages/iou/MoneyRequestAmountForm.tsx:60]: @s77rt do you mean we won't pass `currentCurrency` from `MoneyRequestAmountInput.tsx` but use `moneyRequestAmountInput.current?.getCurrency()` here?

[the remark this answers is not in the document]
Comment (s77rt) [reply src/pages/iou/MoneyRequestAmountForm.tsx:60]: Yes

File: src/components/MoneyRequestAmountInput.tsx (hunk 1 above)
Comment (s77rt) [reply src/components/MoneyRequestAmountInput.tsx:179]: I don't have a strong preference, but this seems like something nice to have

File: src/pages/iou/MoneyRequestAmountForm.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/pages/iou/MoneyRequestAmountForm.tsx:60]: @s77rt I updated.

**Comment by nkdengineer** [comment]
@s77rt I updated please help to check again.

File: src/pages/iou/MoneyRequestAmountForm.tsx [hunk 3]
```diff
@@ -261,7 +262,7 @@ function MoneyRequestAmountForm(
                     amount={amount}
                     currency={currency}
                     isCurrencyPressable={isCurrencyPressable}
-                    onCurrencyButtonPress={onCurrencyButtonPress}
+                    onCurrencyButtonPress={() => onCurrencyButtonPress?.(moneyRequestAmountInput.current?.getCurrency())}
```
Comment (s77rt) [inline src/pages/iou/MoneyRequestAmountForm.tsx:265]: This is not what I meant. We shouldn't pass any prop to `onCurrencyButtonPress`, instead on `IOURequestStepAmount` we should use the input ref's method `getCurrency`

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/pages/iou/MoneyRequestAmountForm.tsx (hunk 3 above)
Comment (nkdengineer) [reply src/pages/iou/MoneyRequestAmountForm.tsx:265]: If we do that we also need to add a prop `moneyRequestAmountInput` as we do in `MoneyRequestAmountInput`.

File: src/components/MoneyRequestAmountInput.tsx [hunk 4]
```diff
@@ -157,22 +162,33 @@ function MoneyRequestAmountInput(
         (newAmount: string) => {
             // Remove spaces from the newAmount value because Safari on iOS adds spaces when pasting a copied value
             // More info: https://github.com/Expensify/App/issues/16974
-            const newAmountWithoutSpaces = MoneyRequestUtils.stripSpacesFromAmount(newAmount);
+            const {currency: updateCurrency, amount: updateAmount} = MoneyRequestUtils.validateAmountWithCurrency(newAmount) ?? {};
```
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx]: I updated.

[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx]: I updated.

File: src/pages/iou/MoneyRequestAmountForm.tsx (hunk 3 above)
Comment (s77rt) [reply src/pages/iou/MoneyRequestAmountForm.tsx:265]: I just realized that we have two refs coming out from `MoneyRequestAmountInput`, that shouldn't happen, let's remove the `moneyRequestAmountInputRef` prop and use the `forwardedRef` only and apply `useImperativeHandle` on it and add missing methods (e.g. focus, blur)

File: src/pages/iou/MoneyRequestAmountForm.tsx [hunk 5]
```diff
@@ -108,7 +105,7 @@ function MoneyRequestAmountForm(
     const isFocused = useIsFocused();
     const wasFocused = usePrevious(isFocused);
 
-    const formattedTaxAmount = CurrencyUtils.convertToDisplayString(Math.abs(taxAmount), currency);
+    const formattedTaxAmount = CurrencyUtils.convertToDisplayString(Math.abs(taxAmount), moneyRequestAmountInput?.current?.getCurrency());
```
Comment (s77rt) [inline src/pages/iou/MoneyRequestAmountForm.tsx:108]: Hmm, looks we are running into a problem, what's the source of truth, the `currency` prop or the `getCurrency` method?

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 6]
```diff
@@ -102,14 +103,14 @@ function IOURequestStepAmount({
...
-            focusTimeoutRef.current = setTimeout(() => textInput.current?.focus(), CONST.ANIMATED_TRANSITION);
+            focusTimeoutRef.current = setTimeout(() => moneyRequestAmountInput?.current?.focus(), CONST.ANIMATED_TRANSITION);
             return () => {
                 if (!focusTimeoutRef.current) {
                     return;
                 }
                 clearTimeout(focusTimeoutRef.current);
             };
-        }, []),
+        }, [moneyRequestAmountInput]),
```
Comment (s77rt) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:113]: Why is this needed as a dependency? Refs do not change

**Review by s77rt — CHANGES_REQUESTED** [review]

**Comment by s77rt** [comment]
Bug: Shouldn't be able to change currency if only a part of the amount is being changed


https://github.com/user-attachments/assets/bda074cc-0318-450a-95a8-c09e0e77b0a9

**Comment by s77rt** [comment]
Bug: Can't switch back to original currency after changing it via paste


https://github.com/user-attachments/assets/4d0089f5-48e3-44d8-82d8-b9083fcf06f8

File: src/components/MoneyRequestAmountInput.tsx [hunk 7]
```diff
@@ -129,16 +132,20 @@ function MoneyRequestAmountInput(
...
-    forwardedRef: ForwardedRef<BaseTextInputRef>,
+    forwardedRef: ForwardedRef<MoneyRequestAmountInputRef>,
 ) {
     const {toLocaleDigit, numberFormat} = useLocalize();
 
     const textInput = useRef<BaseTextInputRef | null>(null);
 
-    const decimals = CurrencyUtils.getCurrencyDecimals(currency);
-    const selectedAmountAsString = amount ? onFormatAmount(amount, currency) : '';
+    const selectedAmountAsString = amountFromProps ? onFormatAmount(amountFromProps, currencyFromProps) : '';
```
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx:141]: `selectedAmountAsString` is the initial value that we use to initial the value for `amout` and `selection` state.

File: src/components/MoneyRequestAmountInput.tsx [hunk 8]
```diff
@@ -260,18 +292,18 @@ function MoneyRequestAmountInput(
         if (!formatAmountOnBlur) {
             return;
         }
-        const formattedAmount = onFormatAmount(amount, currency);
+        const formattedAmount = onFormatAmount(amountFromProps, currencyFromProps);
```
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx:295]: Currently, this logic in the past is amount prop.

File: src/pages/iou/MoneyRequestAmountForm.tsx (hunk 5 above)
Comment (nkdengineer) [reply src/pages/iou/MoneyRequestAmountForm.tsx:108]: `getCurrency` method, I see this problem when we diplsay the limit error for tax amount.

File: src/pages/iou/request/step/IOURequestStepAmount.tsx (hunk 6 above)
Comment (nkdengineer) [reply src/pages/iou/request/step/IOURequestStepAmount.tsx:113]: When I remove this, the `react-complier` check fails.

File: src/components/MoneyRequestAmountInput.tsx [hunk 9]
```diff
@@ -154,25 +161,39 @@ function MoneyRequestAmountInput(
...
     const setNewAmount = useCallback(
-        (newAmount: string) => {
+        (newAmountInput: string) => {
+            const [newCurrencySymbol, newAmount = ''] = newAmountInput.split(/(\d.*)/, 2);
+            const newCurrency = CurrencyUtils.getCurrencyCode(newCurrencySymbol);
+
+            // Return early if the amount is empty or the currency symbol is invalid
+            if (newCurrencySymbol && !newCurrency) {
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx:171]: We only pass the amount string without currency string to `validateAmount` checks.

**Comment by nkdengineer** [comment]
> Bug: Shouldn't be able to change currency if only a part of the amount is being changed

@s77rt  So if we paste the amount with currency for only a part of the amount, should we update the amount or not

> Bug: Can't switch back to original currency after changing it via paste

The problem is when we open the currency page the selected currency is the currency param in the URL and when we choose the original currency, `currencyFromProps` isn't changed then `currency` prop can't be changed.

File: src/components/MoneyRequestAmountInput.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (s77rt) [reply src/components/MoneyRequestAmountInput.tsx:141]: Correct but we should use the state value being the source of truth. In the initial render it will point to the prop value as well

File: src/components/MoneyRequestAmountInput.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (s77rt) [reply src/components/MoneyRequestAmountInput.tsx:171]: I meant to include the condition in the if statement
```js
const isInvalidCurrency = newCurrencySymbol && !newCurrency;
const isInvalidAmount = MoneyRequestUtils.validateAmount(finalAmount, decimals);
if (isInvalidCurrency || isInvalidAmount) {
    setSelection((prevSelection) => ({...prevSelection}));
    return;
}
```

File: src/components/MoneyRequestAmountInput.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (s77rt) [reply src/components/MoneyRequestAmountInput.tsx:295]: That does not look correct, can we use the state value instead?

File: src/pages/iou/MoneyRequestAmountForm.tsx (hunk 5 above)
Comment (s77rt) [reply src/pages/iou/MoneyRequestAmountForm.tsx:108]: How can we make this easier to understand? Most importantly when to use `currency` vs `getCurrency`

**Comment by s77rt** [comment]
@nkdengineer 

> So if we paste the amount with currency for only a part of the amount, should we update the amount or not

No, I think it's safer to consider this as invalid change

File: src/pages/iou/MoneyRequestAmountForm.tsx (hunk 5 above)
Comment (nkdengineer) [reply src/pages/iou/MoneyRequestAmountForm.tsx:108]: The bug happens when we paste another currency and then enter the tax amount that is over limit then the error displays with the old currency if we use `currency`.

What do you think about renaming this to `currencyFromProp`?

File: src/components/MoneyRequestAmountInput.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx:295]: This logic is implemented [here](https://github.com/Expensify/App/pull/40386/files#diff-29ef0426ea09b0af3420c95911645adcd54c0d3d4c0d8c4fa5aa99dfe37a3646) to format the amount after we blur the input on split confirmation flow.

**Comment by nkdengineer** [comment]
@s77rt I fixed two bug above.

File: src/pages/iou/MoneyRequestAmountForm.tsx (hunk 5 above)
Comment (s77rt) [reply src/pages/iou/MoneyRequestAmountForm.tsx:108]: I think it's still unclear when ti use one vs the other. e.g. in `initializeAmount` shouldn't we use the currency from the input? (the function is called on tab switch)

File: src/components/MoneyRequestAmountInput.tsx [hunk 10]
```diff
@@ -180,7 +175,7 @@ function MoneyRequestAmountInput(
 
             // Use a shallow copy of selection to trigger setSelection
             // More info: https://github.com/Expensify/App/issues/16385
-            if (!MoneyRequestUtils.validateAmount(finalAmount, newDecimals)) {
+            if (!MoneyRequestUtils.validateAmount(finalAmount, newDecimals) || (newCurrencySymbol && (!newCurrency || selection.end - selection.start < amount.length - 1))) {
```
Comment (s77rt) [inline src/components/MoneyRequestAmountInput.tsx:178]: break this into separate variables for clarity

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 11]
```diff
@@ -136,6 +136,9 @@ function IOURequestStepAmount({
     };
 
     const navigateToCurrencySelectionPage = () => {
+        if (moneyRequestAmountInput?.current?.getCurrency() !== currency && moneyRequestAmountInput?.current?.getCurrency()) {
+            IOU.setMoneyRequestCurrency(transactionID, moneyRequestAmountInput?.current?.getCurrency(), action === CONST.IOU.ACTION.EDIT);
+        }
```
Comment (s77rt) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:141]: This feels hacky. On navigation we should just navigate without doing side effect (changing currency). The currency should be changed when in fact we change it

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/pages/iou/request/step/IOURequestStepAmount.tsx (hunk 11 above)
Comment (nkdengineer) [reply src/pages/iou/request/step/IOURequestStepAmount.tsx:141]: > The problem is when we open the currency page the selected currency is the currency param in the URL and when we choose the original currency, currencyFromProps isn't changed then currency prop can't be changed.

@s77rt As I mentioned above, the problem is here. Do you have any other ideas for this case?

File: src/components/MoneyRequestAmountInput.tsx [hunk 12]
```diff
@@ -154,25 +161,37 @@ function MoneyRequestAmountInput(
...
                 ? MoneyRequestUtils.stripCommaFromAmount(newAmountWithoutSpaces)
                 : MoneyRequestUtils.replaceCommasWithPeriod(newAmountWithoutSpaces);
+
+            // We should skip updating the amount if the currency symbol is invalid or we paste the amount with currency for only a part of amount
+            const shouldSkip = newCurrencySymbol && (!newCurrency || selection.end - selection.start < amount.length - 1);
+
             // Use a shallow copy of selection to trigger setSelection
             // More info: https://github.com/Expensify/App/issues/16385
-            if (!MoneyRequestUtils.validateAmount(finalAmount, decimals)) {
+            if (!MoneyRequestUtils.validateAmount(finalAmount, newDecimals) || shouldSkip) {
```
Comment (s77rt) [inline src/components/MoneyRequestAmountInput.tsx:181]: Please rename the variable names as in https://github.com/Expensify/App/pull/47767#discussion_r1734950082

File: src/pages/iou/request/step/IOURequestStepAmount.tsx (hunk 11 above)
Comment (s77rt) [reply src/pages/iou/request/step/IOURequestStepAmount.tsx:141]: We shouldn't change the currency until we press the save button that will actually change the currency. Otherwise we will have bugs e.g.
1. Start money request flow with currency A and go to the edit amount page
2. Paste currency B
3. Press the currency button
4. Go back (cancel the edit process)

The currency should be still A

**Comment by s77rt** [comment]
@nkdengineer Any updates here?

**Comment by nkdengineer** [comment]
I will give an update tomorrow.

File: src/components/MoneyRequestAmountInput.tsx (hunk 12 above)
Comment (nkdengineer) [reply src/components/MoneyRequestAmountInput.tsx]: I updated.

**Comment by nkdengineer** [comment]
> Bug: Can't switch back to original currency after changing it via paste

@s77rt For this case, what do you think if we emit an event and pass the selected currency as data if we confirm select a currency in `IOURequestStepCurrency` and then add a listener in `MoneyRequestInput` to update the currency accordingly?

**Comment by s77rt** [comment]
@nkdengineer This adds more complexity, let's try avoid that and look for easier and more straightforward solutions

**Comment by nkdengineer** [comment]
Let me check again for another solution.

**Comment by nkdengineer** [comment]
Still investigating will give an update tomorrow.

**Comment by nkdengineer** [comment]
@s77rt I updated to fix the bug above.

**Comment by s77rt** [comment]
@nkdengineer The main problem still remains, we should not change the currency unless the user actually changes it. Let's not add band-aid solutions and focus on removing the hacky solution that introduced the new bug.

If this is becoming more time consuming it may be worth it to find another approach.

**Comment by nkdengineer** [comment]
@s77rt In this case, I used `modifiedCurrency` as the draft currency then we only update the `currency` field of the transaction if we select the currency in the currency page or click on the confirm button. And in the create flow, we will use `transaction.currency` to create the transaction. For the edit flow, we store the temporary currency in the transaction draft data, and if we close the amount page without saving we don't face any other bug.

**Comment by s77rt** [comment]
@nkdengineer We should avoid changing the currency on navigation. We'd keep having bugs if we don't fix the root cause.

Bug: Shown currency is not correct
1. Start money request flow with currency A and go to the edit amount page
2. Paste currency B
3. Go back
4. Go to the edit amount page again
5. Currency should be A


https://github.com/user-attachments/assets/5259fb50-7202-4d9a-a9f1-ab912c3b3d75

**Comment by nkdengineer** [comment]
Thanks, yesterday I tested but doesn't go to the amount page again. Just stop at the step going back to confirmation page. Will find another solution.

**Comment by nkdengineer** [comment]
@s77rt Updated another solution.

File: src/libs/actions/IOU.ts [hunk 13]
```diff
@@ -406,7 +406,7 @@ function setMoneyRequestCreated(transactionID: string, created: string, isDraft:
 
 function setMoneyRequestCurrency(transactionID: string, currency: string, isEditing = false) {
     const fieldToUpdate = isEditing ? 'modifiedCurrency' : 'currency';
-    Onyx.merge(`${ONYXKEYS.COLLECTION.TRANSACTION_DRAFT}${transactionID}`, {[fieldToUpdate]: currency});
+    Onyx.merge(`${ONYXKEYS.COLLECTION.TRANSACTION_DRAFT}${transactionID}`, {[fieldToUpdate]: currency, currencyUpdatedTime: new Date().getTime()});
```
Comment (s77rt) [inline src/libs/actions/IOU.ts:409]: Using a date here is hacky. And another concern is that once this value is set it won't be unset making it useless and the conditions that check it will always be true which could lead to bugs

**Comment by s77rt** [comment]
Maybe we are overlooking something here? Why we are not facing such bugs with the amount and only with the currency or is it only due to the fact the the currency has a currency selector page which is adding more complexity.

**Comment by nkdengineer** [comment]
@s77rt The problem here is we can change the currency via state and another page. For amount we only change this state in the page.

> Why we are not facing such bugs with the amount and only with the currency or is it only due to the fact the the currency has a currency selector page which is adding more complexity.

**Comment by nkdengineer** [comment]
WHen we go to the currency page, the currency state is passed in the URL then this page display this currency but the data in Onyx is not the same then if we select the original currency, nothing is changed then we can't update the currency state. So the main idea that I resolve this problem here is we need to make the amount page know that we just change the currency then update the currency state.

**Comment by s77rt** [comment]
@nkdengineer This still does not answer the concern highlighted in https://github.com/Expensify/App/pull/47767#discussion_r1755105926. Do you think we can handle this following another approach?

File: src/libs/actions/IOU.ts (hunk 13 above)
Comment (nkdengineer) [reply src/libs/actions/IOU.ts:409]: @s77rt  This will not be a problem because it's updated with `currency`.

Comment (nkdengineer) [reply src/libs/actions/IOU.ts:409]: if we want to reset this value we can reset it in `initMoneyRequest` function.

**Comment by s77rt** [comment]
Bug: Still not able to reselect initial currency

1. Start money request flow with currency A and go to the edit amount page
2. Paste currency B
3. Go to the currency selector page
4. Select currency A
5. Currency should be A


https://github.com/user-attachments/assets/32752087-fc5c-4cfe-8ff2-963e3b55a6ee

**Comment by nkdengineer** [comment]
> Bug: Still not able to reselect initial currency

@s77rt  The problem is if we edit the currency from the confirmation page, we will not call `setMoneyRequestCurreny`. In the second though, I think this solution is simplest and it also fix for the `IOURequestStepTaxAmount` page. What do you think?

> what do you think if we emit an event and pass the selected currency as data if we confirm select a currency in IOURequestStepCurrency and then add a listener in MoneyRequestInput to update the currency accordingly?

**Comment by s77rt** [comment]
@nkdengineer It's been almost a month and we keep running into bugs. I think it's be better to close the PR. Afterall this is a feature request and with the current workflow this feature integration is more complex than initially thought.

**Comment by nkdengineer** [comment]
@s77rt Agree, let's close this PR.
